src/Common/src/System/Security/Cryptography/PasswordBasedEncryption.cs

```diff
@@ -0,0 +1,1032 @@
+                            0);
+
+                        rentedDestination.AsSpan(0, writeOffset).CopyTo(destination);
+
+                        byte[] tmpEnd = decryptor.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
+
+                        fixed (byte* tmpEndPtr = tmpEnd)
+                        {
+                            Span<byte> tmpEndSpan = tmpEnd.AsSpan();
+                            tmpEndSpan.CopyTo(destination.Slice(writeOffset));
+                            CryptographicOperations.ZeroMemory(tmpEndSpan);
+                        }
+
+                        return writeOffset + tmpEnd.Length;
+                    }
+                }
+                finally
+                {
+                    CryptographicOperations.ZeroMemory(tmpKey);
+                    CryptographicOperations.ZeroMemory(tmpIv);
+                    CryptographicOperations.ZeroMemory(rentedEncryptedData.AsSpan(0, encryptedData.Length));
+                    CryptographicOperations.ZeroMemory(rentedDestination.AsSpan(0, destination.Length));
+
+                    ArrayPool.Return(rentedEncryptedData);
+                    ArrayPool.Return(rentedDestination);
+                }
+            }
+        }
+
+        private static void Pbkdf1(
+            IncrementalHash hasher,
+            ReadOnlySpan<byte> password,
+            ReadOnlySpan<byte> salt,
+            int iterationCount,
+            Span<byte> dk)
+        {
+            // The only two hashes that will call into this implementation are
+            // MD5 (16 bytes) and SHA-1 (20 bytes).
+            Span<byte> t = stackalloc byte[20];
+
+            // https://tools.ietf.org/html/rfc2898#section-5.1
+
+            // T_1 = Hash(P || S)
+            hasher.AppendData(password);
+            hasher.AppendData(salt);
+
+            if (!hasher.TryGetHashAndReset(t, out int tLength))
+            {
+                Debug.Fail("TryGetHashAndReset failed with pre-allocated input");
+                throw new CryptographicException();
+            }
+
+            t = t.Slice(0, tLength);
+
+            // T_i = H(T_(i-1))
+            for (int i = 1; i < iterationCount; i++)
+            {
+                hasher.AppendData(t);
+
+                if (!hasher.TryGetHashAndReset(t, out tLength) || tLength != t.Length)
+                {
+                    Debug.Fail("TryGetHashAndReset failed with pre-allocated input");
+                    throw new CryptographicException();
+                }
+            }
+
+            // DK = T_c<0..dkLen-1>
+            t.Slice(0, dk.Length).CopyTo(dk);
+            CryptographicOperations.ZeroMemory(t);
+        }
+
+        internal static void WritePbeAlgorithmIdentifier(
+            AsnWriter writer,
+            bool isPkcs12,
+            string encryptionAlgorithmOid,
+            Span<byte> salt,
+            int iterationCount,
+            string hmacOid,
+            Span<byte> iv)
+        {
+            writer.PushSequence();
```

Comment:
is it possible to use serializer in here?

Reply:
It'd be possible, but because of the inside-out way things would need to be describe that would require a lot of stacked usings, and a couple intermediate byte arrays and probably some allocated Oid objects when only the string was needed.

```C#
if (isPkcs12)
{
    using (AsnWriter pkcs12PbeParams = AsnSerializer.Serialize(
        new PBEParameter
        {
            Salt = salt,
            IterationCount = iterationCount,
        })
    {
        AsnSerializer.Serialize(
            new AlgorithmIdentifierAsn
            {
                Id = new Oid(encryptionAlgorithmOid),
                Parameters = pkcs12PbeParameters.Encode()
            },
            writer);
    }
}
else
{
    // Like above, but with a different using statement for everything after WriteObjectIdentifier
}
```